File: sdk/python/tempmail_sdk/providers/catchmail.py
"""
Catchmail 渠道 — https://catchmail.io
"""

import random
import re
import string
from typing import Any, Dict, List, Optional
from urllib.parse import quote

from .. import http as tm_http
from ..normalize import normalize_email
from ..types import Email, EmailInfo

CHANNEL = "catchmail"
API_BASE = "https://api.catchmail.io/api/v1"
DEFAULT_DOMAIN = "catchmail.io"
DOMAINS = ["catchmail.io", "mailistry.com", "zeppost.com"]

HEADERS = {
    "Accept": "application/json",
    "Referer": "https://catchmail.io/",
    "Origin": "https://catchmail.io",
    "User-Agent": (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
        "(KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0"
    ),
}


def _random_local() -> str:
    chars = string.ascii_lowercase + string.digits
    return "sdk" + "".join(random.choice(chars) for _ in range(16))


def _pick_domain(preferred: Optional[str]) -> str:
    p = (preferred or "").strip().lstrip("@").lower()
    if p:
        for domain in DOMAINS:
            if domain.lower() == p:
                return domain
    return DEFAULT_DOMAIN


def _clean_address(value: Any) -> str:
    if isinstance(value, list):
        return _clean_address(value[0]) if value else ""
    text = str(value or "").strip()
    match = re.search(r"<([^>]+)>", text)
    return match.group(1).strip() if match else text


def _flatten_detail(raw: Dict[str, Any], recipient: str) -> Dict[str, Any]:
    body = raw.get("body")
    if not isinstance(body, dict):
        body = {}
    return {
        "id": raw.get("id", ""),
        "from": _clean_address(raw.get("from")),
        "to": _clean_address(raw.get("to")) or recipient,
        "subject": raw.get("subject", ""),
        "text": body.get("text", ""),
        "html": body.get("html", ""),
        "date": raw.get("date", ""),
        "attachments": raw.get("attachments"),
    }


def _fetch_message(message_id: str, email: str) -> Optional[Dict[str, Any]]:
    resp = tm_http.get(
        f"{API_BASE}/message/{quote(message_id, safe='')}?mailbox={quote(email, safe='')}",
        headers=HEADERS,
        timeout=15,
    )
    if not resp.ok:
        return None
    data = resp.json()
    return data if isinstance(data, dict) else None


def generate_email(domain: Optional[str] = None) -> EmailInfo:
    email = f"{_random_local()}@{_pick_domain(domain)}"
    resp = tm_http.get(
        f"{API_BASE}/mailbox?address={quote(email, safe='')}",
        headers=HEADERS,
        timeout=15,
    )
    resp.raise_for_status()
    return EmailInfo(channel=CHANNEL, email=email)


def get_emails(email: str) -> List[Email]:
    address = (email or "").strip()
    if not address:
        raise ValueError("catchmail: empty email")

    resp = tm_http.get(
        f"{API_BASE}/mailbox?address={quote(address, safe='')}",
        headers=HEADERS,
        timeout=15,
    )
    resp.raise_for_status()
    data = resp.json()
    rows = data.get("messages") if isinstance(data, dict) else []
    if not isinstance(rows, list):
        return []

    emails: List[Email] = []
    for item in rows:
        if not isinstance(item, dict):
            continue
        message_id = str(item.get("id") or "").strip()
        if not message_id:
            continue
        detail = _fetch_message(message_id, address)
        if detail:
            emails.append(normalize_email(_flatten_detail(detail, address), address))
            continue
        emails.append(normalize_email({
            "id": message_id,
            "from": _clean_address(item.get("from")),
            "to": item.get("mailbox") or address,
            "subject": item.get("subject", ""),
            "date": item.get("date", ""),
        }, address))
    return emails

File: sdk/ruby/lib/tempmail_sdk/providers/catchmail.rb
# frozen_string_literal: true

require "uri"

module TempmailSdk
  module Providers
    # Catchmail 渠道 — https://catchmail.io
    # 匿名 REST API：创建邮箱即列表接口，逐封拉取详情
    module Catchmail
      CHANNEL = "catchmail"
      API_BASE = "https://api.catchmail.io/api/v1"
      DEFAULT_DOMAIN = "catchmail.io"
      DOMAINS = %w[catchmail.io mailistry.com zeppost.com].freeze
      HEADERS = {
        "Accept" => "application/json",
        "Referer" => "https://catchmail.io/",
        "Origin" => "https://catchmail.io",
        "User-Agent" => "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0"
      }.freeze

      module_function

      def random_local
        chars = ("a".."z").to_a + ("0".."9").to_a
        "sdk#{Array.new(16) { chars.sample }.join}"
      end

      # 选择域名：命中别名域名则用之，否则默认域名
      def pick_domain(preferred)
        p = preferred.to_s.strip.sub(/\A@/, "").downcase
        unless p.empty?
          found = DOMAINS.find { |d| d.downcase == p }
          return found if found
        end
        DEFAULT_DOMAIN
      end

      # 从可能带显示名的地址字段提取纯地址
      def clean_address(value)
        value = value.first if value.is_a?(Array)
        text = value.to_s.strip
        m = text.match(/<([^>]+)>/)
        m ? m[1].strip : text
      end

      def flatten_detail(raw, recipient)
        body = raw["body"].is_a?(Hash) ? raw["body"] : {}
        {
          "id" => raw["id"] || "", "from" => clean_address(raw["from"]),
          "to" => (a = clean_address(raw["to"])).empty? ? recipient : a,
          "subject" => raw["subject"] || "", "text" => body["text"] || "",
          "html" => body["html"] || "", "date" => raw["date"] || "",
          "attachments" => raw["attachments"]
        }
      end

      def fetch_message(message_id, email)
        resp = Http.get(
          "#{API_BASE}/message/#{URI.encode_www_form_component(message_id)}?mailbox=#{URI.encode_www_form_component(email)}",
          headers: HEADERS, timeout: 15
        )
        return nil unless resp.ok?

        data = resp.json
        data.is_a?(Hash) ? data : nil
      end

      # @param domain [String, nil]
      # @return [EmailInfo]
      def generate_email(domain = nil)
        email = "#{random_local}@#{pick_domain(domain)}"
        Http.get("#{API_BASE}/mailbox?address=#{URI.encode_www_form_component(email)}",
                 headers: HEADERS, timeout: 15).raise_for_status
        EmailInfo.new(channel: CHANNEL, email: email)
      end

      # @param email [String]
      # @return [Array<Email>]
      def get_emails(email)
        address = email.to_s.strip
        raise "catchmail: empty email" if address.empty?

        resp = Http.get("#{API_BASE}/mailbox?address=#{URI.encode_www_form_component(address)}",
                        headers: HEADERS, timeout: 15)
        resp.raise_for_status
        data = resp.json
        rows = data.is_a?(Hash) ? data["messages"] : nil
        return [] unless rows.is_a?(Array)

        rows.filter_map do |item|
          next unless item.is_a?(Hash)

          message_id = item["id"].to_s.strip
          next if message_id.empty?

          detail = fetch_message(message_id, address)
          if detail
            Normalize.normalize_email(flatten_detail(detail, address), address)
          else
            Normalize.normalize_email({
                                        "id" => message_id, "from" => clean_address(item["from"]),
                                        "to" => item["mailbox"] || address,
                                        "subject" => item["subject"] || "", "date" => item["date"] || ""
                                      }, address)
          end
        end
      end
    end
  end
end
